File: src/GnOuGo.Flow.Core/Runtime/Executors/LoopSequentialExecutor.cs
using System.Text.Json.Nodes;
using GnOuGo.Flow.Core.Expressions;
using GnOuGo.Flow.Core.Models;

namespace GnOuGo.Flow.Core.Runtime.Executors;

/// <summary>
/// Sequential loop: iterates over an <c>items</c> (or <c>over</c>) array, or loops <c>times</c>/<c>while</c> times.
/// </summary>
public sealed class LoopSequentialExecutor : IStepExecutor
{
    public string StepType => "loop.sequential";

    public IReadOnlyList<StepExceptionDoc>? DocumentedExceptions => new StepExceptionDoc[]
    {
        new(ErrorCodes.InputValidation, false, "The step is missing its nested `steps` collection or has invalid loop input values."),
        new(ErrorCodes.LoopLimit, false, "The loop exceeded `max_times` or the runtime maximum loop iteration limit.")
    };

    public string DslSnippet => """
        ### loop.sequential — Sequential loop over items or with while/times
        ```yaml
        # Iterate over a list sequentially:
        - id: process
          type: loop.sequential
          input:
            items: "${data.inputs.my_list}"    # required — array expression (alias: 'over')
          item_var: item                        # variable name for current item (default: "item")
          index_var: idx                        # variable name for current index (default: "i")
          steps:
            - id: transform
              type: set
              input: { value: "${data.item}" }

        # Fixed iteration count:
        - id: loop
          type: loop.sequential
          input:
            times: 5
            # or: while: "${data._loop.index < 10}"
          steps:
            - id: iter
              type: template.render
              input: { engine: mustache, template: "Iteration {{idx}}", data: { idx: "${data._loop.index}" }, mode: text }
        ```
        Context: `data.<item_var>` (current item), `data.<index_var>` (current index), `data._loop.index`, `data._loop.item` (when iterating items).
        Output: `{ results: [...], count: N }`
        """;

    public async Task<JsonNode?> ExecuteAsync(StepExecutionContext ctx, CancellationToken ct)
    {
        var subSteps = ctx.Step.Steps
            ?? throw new WorkflowRuntimeException(ErrorCodes.InputValidation, "loop.sequential requires 'steps'");

        var input = ctx.Engine.GetResolvedInput(ctx);
        var inputObj = input as JsonObject;

        int maxTimes = ctx.Limits.MaxLoopIterations;
        if (inputObj?.TryGetPropertyValue("max_times", out var mt) == true && mt != null)
            maxTimes = (int)ExpressionEvaluator.GetNumber(mt);

        // Resolve items array: support both 'items' and 'over' (alias)
        JsonArray? items = null;
        bool itemsKeyPresent = false;
        JsonNode? rawItemsValue = null;
        if (inputObj?.TryGetPropertyValue("items", out var itemsNode) == true && itemsNode != null)
        {
            itemsKeyPresent = true;
            rawItemsValue = itemsNode;
            if (itemsNode is JsonArray itemsArr)
                items = itemsArr;
        }
        else if (inputObj?.TryGetPropertyValue("over", out var overNode) == true && overNode != null)
        {
            itemsKeyPresent = true;
            rawItemsValue = overNode;
            if (overNode is JsonArray overArr)
                items = overArr;
        }

        int? times = null;
        if (inputObj?.TryGetPropertyValue("times", out var t) == true && t != null)
            times = (int)ExpressionEvaluator.GetNumber(t);

        // Validate: items and times are mutually exclusive
        if (items != null && times != null)
            throw new WorkflowRuntimeException(ErrorCodes.InputValidation,
                "loop.sequential: 'items'/'over' and 'times' are mutually exclusive");

        // Validate: if 'items'/'over' key was specified but didn't resolve to an array, fail early
        if (itemsKeyPresent && items == null)
            throw new WorkflowRuntimeException(ErrorCodes.InputValidation,
                "loop.sequential: 'items'/'over' must resolve to an array. "
                + $"Got: {rawItemsValue?.GetValueKind().ToString() ?? "null"}. "
                + "Check that the expression returns a JSON array.");

        // ── items-based iteration ──────────────────────────────────────
        if (items != null)
        {
            var itemVar = ctx.Step.Source.ItemVar ?? "item";
            var indexVar = ctx.Step.Source.IndexVar ?? "i";

            if (items.Count > maxTimes)
                throw new WorkflowRuntimeException(ErrorCodes.LoopLimit,
                    $"Loop items ({items.Count}) exceeds limit ({maxTimes})");

            var iterations = new JsonArray();

            for (int i = 0; i < items.Count; i++)
            {
                ct.ThrowIfCancellationRequested();

                var currentItem = items[i]?.DeepClone();
                ctx.Data[itemVar] = currentItem?.DeepClone();
                ctx.Data[indexVar] = JsonValue.Create(i);
                ctx.Data["_loop"] = new JsonObject { ["index"] = JsonValue.Create(i), ["item"] = currentItem?.DeepClone() };
                ctx.Data["loop"] = new JsonObject { ["index"] = JsonValue.Create(i), ["item"] = currentItem?.DeepClone() };

                // Evaluate while condition if present (combined items + while)
                if (inputObj?.TryGetPropertyValue("while", out var whileExpr) == true && whileExpr != null)
                {
                    var whileStr = ExpressionEvaluator.GetString(whileExpr);
                    var condResult = ctx.Interpolator.Interpolate(whileStr, ctx.Data);
                    if (!ExpressionEvaluator.GetBool(condResult))
                        break;
                }

                var result = new RunResult { Success = true };
                await ctx.Engine.ExecuteStepsAsync(
                    subSteps,
                    ctx.Data,
                    result,
                    ctx.Limits,
                    ctx.CallDepth,
                    ctx.CallStack,
                    ctx.EffectiveExecutionScope,
                    ct,
                    ctx.TelemetrySpan);

                iterations.Add(ctx.Data["steps"]?.DeepClone() ?? new JsonObject());
            }

            // Clean up loop-scoped variables
            ctx.Data.Remove(itemVar);
            ctx.Data.Remove(indexVar);
            ctx.Data.Remove("_loop");
            ctx.Data.Remove("loop");

            return new JsonObject { ["results"] = iterations, ["count"] = iterations.Count };
        }

        // ── times / while iteration ────────────────────────────────────
        bool hasWhile = inputObj?.ContainsKey("while") == true;

        // Guard: at least one termination condition must be present
        if (!times.HasValue && !hasWhile)
            throw new WorkflowRuntimeException(ErrorCodes.InputValidation,
                "loop.sequential requires at least one of: 'items', 'over', 'times', or 'while'");

        var whileIterations = new JsonArray();
        int iteration = 0;

        while (true)
        {
            ct.ThrowIfCancellationRequested();

            if (times.HasValue && iteration >= times.Value)
                break;

            if (iteration >= maxTimes)
                throw new WorkflowRuntimeException(ErrorCodes.LoopLimit,
                    $"Loop iteration limit reached ({maxTimes})");

            ctx.Data["_loop"] = new JsonObject { ["index"] = iteration };
            ctx.Data["loop"] = new JsonObject { ["index"] = iteration };

            // Evaluate while condition
            if (inputObj?.TryGetPropertyValue("while", out var whileExpr2) == true && whileExpr2 != null)
            {
                var whileStr = ExpressionEvaluator.GetString(whileExpr2);
                var condResult = ctx.Interpolator.Interpolate(whileStr, ctx.Data);
                if (!ExpressionEvaluator.GetBool(condResult))
                    break;
            }

            // Execute iteration
            var result2 = new RunResult { Success = true };
            await ctx.Engine.ExecuteStepsAsync(
                subSteps,
                ctx.Data,
                result2,
                ctx.Limits,
                ctx.CallDepth,
                ctx.CallStack,
                ctx.EffectiveExecutionScope,
                ct,
                ctx.TelemetrySpan);

            whileIterations.Add(ctx.Data["steps"]?.DeepClone() ?? new JsonObject());
            iteration++;
        }

        return new JsonObject { ["results"] = whileIterations, ["count"] = iteration };
    }
}

File: librairies/python/gnougo-flow-core/src/gnougo_flow_core/runtime_steps/loop_sequential_executor.py
from __future__ import annotations

import copy

from gnougo_flow_core.runtime import *  # noqa: F401,F403


class LoopSequentialExecutor:
    step_type = "loop.sequential"
    step_description = "Sequential loop: iterates over an items/over array, or loops times/while times."
    dsl_snippet = """
### loop.sequential — Sequential loop over items or with while/times
```yaml
# Iterate over a list sequentially:
- id: process
  type: loop.sequential
  input:
    items: "${data.inputs.my_list}"    # required — array expression (alias: 'over')
  item_var: item                        # variable name for current item (default: "item")
  index_var: idx                        # variable name for current index (default: "i")
  steps:
    - id: transform
      type: set
      input: { value: "${data.item}" }

# Fixed iteration count:
- id: loop
  type: loop.sequential
  input:
    times: 5
    # or: while: "${data._loop.index < 10}"
  steps:
    - id: iter
      type: template.render
      input: { engine: mustache, template: "Iteration {{idx}}", data: { idx: "${data._loop.index}" }, mode: text }
```
Context: `data.<item_var>` (current item), `data.<index_var>` (current index), `data._loop.index`, `data._loop.item` (when iterating items).
Output: `{ iterations: [...], count: <number> }`.
"""
    documented_exceptions = [
        (ErrorCodes.INPUT_VALIDATION, False, "loop.sequential requires nested 'steps' or has incompatible inputs."),
        (ErrorCodes.LOOP_LIMIT, False, "loop.sequential exceeded max_times limit."),
    ]

    async def execute_async(self, ctx: StepExecutionContext) -> Any:
        if not ctx.step.steps:
            raise WorkflowRuntimeException(ErrorCodes.INPUT_VALIDATION, "loop.sequential requires 'steps'")
        input_obj = ctx.engine.get_resolved_input(ctx) or {}
        if not isinstance(input_obj, dict):
            input_obj = {}

        times = input_obj.get("times")
        # Support both 'items' (primary) and 'over' (alias) for array iteration
        items = input_obj.get("items") or input_obj.get("over")
        while_expr = input_obj.get("while")
        max_times = int(input_obj.get("max_times", ctx.limits.max_loop_iterations))

        # Validate: items and times are mutually exclusive
        if items is not None and times is not None:
            raise WorkflowRuntimeException(
                ErrorCodes.INPUT_VALIDATION,
                "loop.sequential: 'items'/'over' and 'times' are mutually exclusive",
            )

        if items is not None and not isinstance(items, list):
            raise WorkflowRuntimeException(
                ErrorCodes.INPUT_VALIDATION,
                "loop.sequential 'items'/'over' must be an array",
            )

        item_var = ctx.step.source.item_var or "item"
        index_var = ctx.step.source.index_var or "i"
        iterations: list[Any] = []

        # ── items-based iteration ──────────────────────────────────────
        if items is not None:
            if len(items) > max_times:
                raise WorkflowRuntimeException(
                    ErrorCodes.LOOP_LIMIT,
                    f"Loop items ({len(items)}) exceeds limit ({max_times})",
                )

            for i, item in enumerate(items):
                ctx.data[item_var] = copy.deepcopy(item)
                ctx.data[index_var] = i
                ctx.data["_loop"] = {"index": i, "item": copy.deepcopy(item)}
                ctx.data["loop"] = {"index": i, "item": copy.deepcopy(item)}

                # Evaluate while condition if present (combined items + while)
                if while_expr is not None:
                    cond = ctx.engine.interpolator.interpolate(str(while_expr), ctx.data)
                    if not ExpressionEvaluator.get_bool(cond):
                        break

                run = RunResult(success=True)
                await ctx.engine.execute_steps_async(
                    ctx.step.steps, ctx.data, run, ctx.limits, ctx.call_depth, ctx.call_stack, ctx.telemetry_span, ct=ctx.ct,
                )
                iterations.append(copy.deepcopy(ctx.data.get("steps", {})))

            # Clean up loop-scoped variables
            ctx.data.pop(item_var, None)
            ctx.data.pop(index_var, None)
            ctx.data.pop("_loop", None)
            ctx.data.pop("loop", None)

            return {"iterations": iterations, "count": len(iterations)}

        # ── times / while iteration ────────────────────────────────────
        i = 0
        while True:
            if times is not None and i >= int(ExpressionEvaluator.get_number(times)):
                break
            if i >= max_times:
                raise WorkflowRuntimeException(
                    ErrorCodes.LOOP_LIMIT,
                    f"Loop iteration limit reached ({max_times})",
                )

            ctx.data["_loop"] = {"index": i}
            ctx.data["loop"] = {"index": i}

            # while: evaluated every iteration BEFORE running steps
            if while_expr is not None:
                cond = ctx.engine.interpolator.interpolate(str(while_expr), ctx.data)
                if not ExpressionEvaluator.get_bool(cond):
                    break

            run = RunResult(success=True)
            await ctx.engine.execute_steps_async(
                ctx.step.steps, ctx.data, run, ctx.limits, ctx.call_depth, ctx.call_stack, ctx.telemetry_span, ct=ctx.ct,
            )
            iterations.append(copy.deepcopy(ctx.data.get("steps", {})))
            i += 1

        return {"iterations": iterations, "count": i}
